src/components/implementation/cbuf_mgr/default/cbuf_mgr.c

```diff
@@ -763,54 +842,222 @@ cbuf_unmap_at(spdid_t s_spd, unsigned int cbid, spdid_t d_spd, vaddr_t d_addr)
+ */
+static void
+__get_nfo(struct cbuf_comp_info *cci)
+{
+	printc("Now iterating over global cbuf list\n");
+	// iterate over cbi_head until done
+	struct cbuf_info *current = cbi_head;
+	if (!current) {
+		printc("There should be at least one cbuf in the system\n");
+		BUG();
+	}
+
+	do {
+		// The cbuf_info list we made is actually crap because it copies the pointers so if the structure is updated it is no longer valid (how to fix???)
+		// instead, re-get the cbi from the cbuf_id which we assume is constant
+		struct cbuf_info *cbi = cmap_lookup(&cbufs, current->cbid);
+
+		if (cbi) { /* presumably the cbuf could have been deleted but this is unlikely */
+			printc("cbuf #%2u with size %x, mem start %x, and maps ", cbi->cbid, cbi->size, cbi->mem);
+			struct cbuf_maps *m = &cbi->owner;
+			struct cbuf_meta *cm = m->m;		/* Use this after loop */
+
+			do {
+				printc("[%d:%x] ", m->spdid, m->addr);
+				m = FIRST_LIST(m, next, prev);
+			} while (m != &cbi->owner);
+
+			printc("\n\tnfo: [ex %d|ow %d|t %d|in %d|re %d] page pointer %x",
+									CBUF_EXACTSZ(cm)      >> 11,
+									CBUF_OWNER(cm)        >> 10,
+									CBUF_TMEM(cm)         >>  9,
+									CBUF_INCONSISTENT(cm) >>  8,
+									CBUF_RELINQ(cm)       >>  7,
+									CBUF_PTR(cm));
+			
+			printc("\n");
+		}
+
+		current = FIRST_LIST(current, next, prev);
+	} while (current != cbi_head);
+
+	printc("Done iterating over global cbuf list\n");
 }
 
-void cos_fix_spdid_metadata(spdid_t o_spd, spdid_t f_spd)
+/* This is internal so the real core of the function doesn't take a lock */
+static vaddr_t
+__cbuf_fork_spd(spdid_t o_spd, spdid_t f_spd, int cinfo_cbid)
 {
-	int r;
+	struct cbuf_comp_info *src, *dst;
+	struct cbuf_info *current;
+	int i;
+	vaddr_t ret = (vaddr_t) NULL;		/* address of cinfo page */
 
-	printd("cbuf: cos_fix_spdid_metadata for %d -> %d\n", o_spd, f_spd);
+	printc("starting valloc_fork_spd from %d to %d\n", o_spd, f_spd);
+	int r = valloc_fork_spd(cos_spd_id(), o_spd, f_spd);
+	printc("finished valloc_fork_spd with ret %d\n", r);
+	
+	src = cbuf_comp_info_get(o_spd);
+	dst = cbuf_comp_info_get(f_spd);
+	__get_nfo(src);
+
+	/* Should create a new shared page between cbuf_mgr and dst */
+	dst->csp = NULL;
+	//dst->dest_csp = src->dest_csp;
+	
+	current = cbi_head;
+	vaddr_t r_addr;
+	do {
+		// The cbuf_info list we made is actually crap because it copies the pointers so if the structure is updated it is no longer valid (how to fix???)
+		// instead, re-get the cbi from the cbuf_id which we assume is constant
+		struct cbuf_info *cbi = cmap_lookup(&cbufs, current->cbid);
+		struct cbuf_maps *m = &cbi->owner;
 	
-	r = cbuf_fork_spd(cos_spd_id(), o_spd, f_spd);
-	if (r) printd("Error (%d) in cbuf_fork_spd\n", r);
-	/* TODO: invoke valloc to get transitive fixup? */
+		if (m->spdid == o_spd) {
+			/* This is for if O is the owner */
+			/* This just universally forks everything to a new cbuf. */
+			__cbuf_fork_cbuf(o_spd, cbi->cbid, f_spd, 0);
```

Comment:
That whole function needs a redesign. Currently it returns 0 on success often but q_addr once we find that. I'll give it some thought.